software/firmware/europi.py

```diff
@@ -1,249 +1,298 @@
+        # Over-samples the ADC and returns the average.
+        values = []
+        for _ in range(samples or self._samples):
+            values.append(self.pin.read_u16())
+        return round(sum(values) / len(values))
+
+    def set_samples(self, samples):
+        """Override the default number of sample reads with the given value."""
+        if not isinstance(samples, int):
+            raise ValueError(
+                f"set_samples expects an int value, got: {samples}")
+        self._samples = samples
 
-    def unit_interval(self, samples=256):
-        return 1 - (sample_adc(self.input, samples) / MAX_UINT12) #Provide the knob's position as a float value between 0.0 and 1.0
+    def percent(self, samples=None):
+        """Return the percentage of the component's current relative range."""
+        return self._sample_adc(samples) / MAX_UINT16
 
-    def read_position(self, steps=100, samples=256): #Returns an int in the range of steps based on knob position.
+    def range(self, steps=100, samples=None):
+        """Return a value (upper bound excluded) chosen by the current voltages relative position."""
         if not isinstance(steps, int):
-            raise Exception("Please only use integer type for steps and decimal_places with the read_position method")
+            raise ValueError(f"range expects an int value, got: {steps}")
+        percent = self.percent(samples)
+        if int(percent) == 1:
+            return steps -1
+        return int(percent * steps)
+
+    def choice(self, values, samples=None):
+        """Return a value from a list chosen by the relative knob position."""
+        if not isinstance(values, list):
+            raise ValueError(f"choice expects a list, got: {values}")
+        percent = self.percent(samples)
+        if percent == 1.0:
+            return values[-1]
+        return values[int(percent * len(values))]
+
+
+class AnalogueInput(AnalogueReader):
+    """A class for handling the reading of analogue control voltage."""
+
+    def __init__(self, pin, min_voltage=MIN_INPUT_VOLTAGE, max_voltage=MAX_INPUT_VOLTAGE):
+        super().__init__(pin)
+        self.MIN_VOLTAGE = min_voltage
+        self.MAX_VOLTAGE = max_voltage
+        self._gradients = []
+        for index, value in enumerate(CALIBRATION_VALUES[:-1]):
+            self._gradients.append(1 / (CALIBRATION_VALUES[index+1] - value))
+        self._gradients.append(self._gradients[-1])
+
+    def percent(self, samples=None):
+        """Current voltage as a relative percentage of the component's range."""
+        # Determine the percent value from the max calibration value.
+        reading = self._sample_adc(samples)
+        max_value = max(reading, CALIBRATION_VALUES[-1])
+        return reading / max_value
+
+    def read_voltage(self, samples=None):
+        reading = self._sample_adc(samples)
+        max_value = max(reading, CALIBRATION_VALUES[-1])
+        percent = reading / max_value

```

Comment:
This could be `percent = self.percent(samples)`

Reply:
ACK. See comments in https://github.com/Allen-Synthesis/EuroPi/pull/33#discussion_r789737175